### ISSUE TITLE
Hijacking API Change

### ISSUE THREAD

**Issue by gustavnikolaj** [gustavnikolaj/hijackresponse#28]
This issue is a continuation of a conversation in #24. Summed up by the quote from @papandreou below.

> I still think it would be worth it to try to get rid of the callback one way or the other. I guess that really would need to happen before 5.0.0, as it'd be pretty douchy to break backwards compatibility twice :)

---

# Current Situation

In v4 highjacking was done like this:

```js
app.use(function(req, res, next) {
  hijackResponse(res, function(err, res) {
    if (err) {
      res.unhijack(); // Make the original res object work again
      return next(err);
    }

    // Don't hijack HTML responses:
    if (/^text\/html(?:;$)/.test(res.getHeader("Content-Type"))) {
      return res.unhijack();
    }

    res.setHeader("X-Hijacked", "yes!");
    res.removeHeader("Content-Length");

    res.pipe(transformStream).pipe(res);
  });
  // next() must be called explicitly, even when hijacking the response:
  next();
});
```

The changes in v5 so far already improves this:

1. You don't get the error argument in your callback anymore. https://github.com/gustavnikolaj/hijackresponse/commit/41c851ad31e9dff92370cfdc56fb4f03fda41d5e 
2. With the split-stream PR merged (#24) you don't need res.unhijack to eject from the highjack, and the overall API is easier and less surprising, as you don't have a res with an extra different stream patched on to it.

```js
app.use(function(req, res, next) {
  hijackResponse(res, function(hijackedResponseBody, res) {
    // Don't hijack HTML responses:
    if (/^text\/html(?:;$)/.test(res.getHeader("Content-Type"))) {
      // forward the hijacked data with no modifications instead of calling the unhijack method
      return hijackedResponseBody.pipe(res);
    }

    res.setHeader("X-Hijacked", "yes!");
    res.removeHeader("Content-Length");

    // almost the same as before, besides res only acting as a writeable now
    hijackedResponseBody.pipe(transformStream).pipe(res);
  });
  // next() must be called explicitly, even when hijacking the response:
  next();
});
```

# Goals

1. Getting rid of the callback

# Circumstances

The artifacts from the `hijackResponse`-function are actually all available synchronously:

1. We monkey patch the passed in `res`.
2. We create a `TransformStream` which the monkey-patched `res` will write into. We pass this stream into the callback as the first argument, and communicate about it as being a `ReadableStream` towards the caller (not in code, only in docs).
3. We create a `proxyRes` object, which is a copy of `res`, but with the original methods put back on. We pass this into the callback as the second argument.

The callback is currently called when the monkey-patched res would attempt to write the headers to the stream.

- It happens implicitly when the first bit of data is written.
- After the headers are sent you can no longer modify status code, status message, or the headers.
- You generally have decided how you want to handle the response at that point, and it will only change if you have an unexpected asynchronous error pop up.

The monkey-patched res will instead of actually calling the original writeHead function call the callback, to handover control to the hijacker.

# Approaches

We need to have a hook called somehow when the writeHead function is called. Unless we are talking very naive hijack-middlewares, you will generally want to modify the headers when you modify the response - altering `Content-Type`, unsetting `Content-Length`, modifying the `ETag` etc.

## Approach 1: Synchronous (INVALID)

If we just made a simple API that monkey patched res, returned the proxyRes and hijackedResponseBodyStream, and didn't provide any timing assistance it would look like this:

```js
app.use(function(req, res, next) {
  const { hijackedReponseBody, proxyRes } = hijackResponse(res);

  if (/^text\/html(?:;$)/.test(res.getHeader("Content-Type"))) {
    // This check will not work, as the middleware serving this file and
    // setting the Content-Type header has not yet run.

    hijackedResponseBody.pipe(proxyRes);
  } else {
    res.setHeader("X-Hijacked", "yes!");
    // This header will be removed, but it will be added back again once the
    // later middlewares are run.
    res.removeHeader("Content-Length");

    hijackedResponseBody.pipe(transformStream).pipe(proxyRes);
  }

  // We still need to call res- we cannot return a pipe to proxyRes, as there
  // will never be an opportunity for people further down to write anything
  // into hijackedResponseBody
  next();
});
```

This example clearly demonstrates the need to run at a certain point in time, when the writers to hijackedResponseBody are done with their request handling.

## Approach 2: Event Listeners

We need to have control handed over at a certain time, and we want to get rid of the callbacks. We know that the headers are implicitly flushed when the first bit of data is written to res. We want our hijacker to get control again when we see the first bit of data written to `hijackedResponseBody`. That sounds a lot like the `readable` event on a `ReadableStream` - or maybe just the first `readable` event.

```js
app.use((req, res, next) => {
  const {hijackedResponseBody, unmonkeyedRes} = hijackResponse(res);

  hijackedResponseBody.once('readable', () => {
    if (/^text\/html(?:;$)/.test(res.getHeader("Content-Type"))) {
        return hijackedResponseBody.pipe(unmonkeyedRes);
    }

    unmonkeyedRes.setHeader("X-Hijacked", "yes!");
    unmonkeyedRes.removeHeader("Content-Length");

    return hijackedResponseBody.pipe(transformStream).pipe(unmonkeyedRes);
  });

  next();
});
```

## Approach 3: Promise based

I need some help on this one - I couldn't get further than this, which is basically just the current callback but called differently.

```js
app.use((req, res, next) => {
    hijackResponse(res);
    res.extractionPoint().then((hijackedResponseBody, res) => {
        // same callback as current situation - different way to set it up.
    }, next);

    next();
})
```

**Comment by gustavnikolaj** [comment]
@papandreou I've tried to enumerate the situation and the different alternatives I could think of. I'd be very happy to have your input.

The only real alternative is ofc Approach 2. You could also use a different event than `readable` - maybe `extraction` (to stay on theme) or `headers`.

The headers almost always need to be modified, or will somehow have influence on what you decide to do with your hijackedResponse (what you are going to pipe it through, returning an error instead, or just passthrough). So it's hard to see a usable interface that doesn't revolve around that callback pattern.

You could maybe make an alternate API that is entirely based on a transform stream, something like:

```js
app.use((req, res, next) => {
  hijackResponse(new TransformStream({
    transform() { /* ... */ }
    flush() { /* ... */ }
  }));
  next();
});
```

But that would be very different from what we are doing now.

**Comment by papandreou** [comment]
What if we handed `next` into `hijackResponse` as well and call it at the right time internally, wouldn't it then be possible to:

```js
const { hijackedReponseBody, proxyRes } = await hijackResponse(res, next);
```

This example ofc assumes that you have a strategy for implementing (some part of) your route using an `async` function, which I know isn't universal. I'm fairly sure it'd work with `express-promise-router` since that still supports calling `next` explicitly:

```js
router = require("express-promise-router")();
 
router.use('/hello', async function (req, res, next) {
  const { hijackedReponseBody, proxyRes } = await hijackResponse(res, next);
  // Do something with the hijacked response...
});
```

But even the more spelled-out promise version of it is kinda nice:

```js
app.use((req, res, next) => {
  hijackResponse(res, next)
    .then(({ hijackedReponseBody, proxyRes }) => { ... });
});
```

**Comment by gustavnikolaj** [comment]
Riffing on your idea, we could just pretend that `proxyRes` was a writable stream that we wired to the original res, and that `hijackedResponseBody` was a readable stream of what others wrote to res (it actually is). `proxyRes` is the restored res object, effectively, and it is actually a writable stream. All methods on  monkey-patched `res` and `proxyRes` except for those that are for the writable-stream are identical.

```js
router.use('/hello', async function (req, res, next) {
  const { readable, writeable } = await hijackResponse(res, next);
  
  if (/^text\/html(?:;$)/.test(res.getHeader("Content-Type"))) {
    return readable.pipe(writeable);
  }

  res.setHeader("X-Hijacked", "yes!");
  res.removeHeader("Content-Length");

  return readable.pipe(transformStream).pipe(writeable);
});
```

This allows transparent reuse of res, as long as you don't use `write`, `send`, `end`, etc...

**Comment by gustavnikolaj** [comment]
If you call `res.write` instead of `writeable.write` you will actually push data into `readable`... :-)

**Comment by gustavnikolaj** [comment]
I tried to implement it in #30. I skipped most of the tests, but the ones that I did convert passed pretty much the first time around :-) It seems really natural, even without the async/await stuff.

**Comment by papandreou** [comment]
That looks super nice! 😍

**Comment by gustavnikolaj** [comment]
Yeah, it actually works out pretty great. Really great idea to pass next in!

Do you have suggestions for the naming of the properties in the returned object from hijackresponse? `readable`/`writeable` may be a little too generic for the names - but `hijackedResponseBody` and `iDontEvenKnowWhatToCallItResponseBody` are a little long... :)

I'll try to polish it up and land updates to the `express-{compiless,processimage,extract-headers}` PRs as soon as I can get some time freed up :)

**Comment by papandreou** [comment]
`readable` and `writeable` would be... okay I guess. I like that it sets up expectations for what you can do with the objects. Maybe `originalBody` / `replacementBody`? Or `originalReadable` / `replacementWritable`. Not great either, idk.

**Comment by gustavnikolaj** [comment]
If you write it out, instead of destructure them, it makes more sense:

```js
const hijackedResponse = await hijackResponse(res, next);

hijackedResponse.readable.pipe(hijackedResponse.writeable);
```

**Comment by gustavnikolaj** [comment]
Fixing the naming of those are less critical - we can always just do it in a minor, and just keep the old ones as getter that warn about an upcoming deprecation ;-)

**Comment by papandreou** [comment]
Ah yeah, good point 👍

**Comment by gustavnikolaj** [comment]
https://github.com/gustavnikolaj/hijackresponse/pull/30 is updated with the entire test suite passing.

**Comment by gustavnikolaj** [comment]
Merged #30 to master.
